src/batchUploader.ts

```diff
@@ -169,47 +207,59 @@ export class BatchUploader {
      * @param triggerFuture whether to trigger the loop again - for manual/forced uploads this should be false
      * @param useBeacon whether to use the beacon API - used when the page is being unloaded
      */
-    private async prepareAndUpload(triggerFuture: boolean, useBeacon: boolean) {
+    private async prepareAndUpload(
+        triggerFuture: boolean,
+        useBeacon: boolean
+    ): Promise<void> {
+        // Fetch current user so that events can be grouped by MPID
         const currentUser = this.mpInstance.Identity.getCurrentUser();
 
-        const currentEvents = this.pendingEvents;
-        this.pendingEvents = [];
-        const newUploads = BatchUploader.createNewUploads(
+        const currentEvents = this.eventsQueuedForProcessing;
+        this.eventsQueuedForProcessing = [];
+
+        const newBatches = BatchUploader.createNewBatches(
             currentEvents,
             currentUser,
             this.mpInstance
         );
-        if (newUploads && newUploads.length) {
-            this.pendingUploads.push(...newUploads);
+
+        if (!isEmpty(newBatches)) {
+            this.batchesQueuedForProcessing.push(...newBatches);
         }
 
-        const currentUploads = this.pendingUploads;
-        this.pendingUploads = [];
-        const remainingUploads: Batch[] = await this.upload(
+        // Clear out pending batches to avoid any potential duplication
+        const batchesToUpload = this.batchesQueuedForProcessing;
+        this.batchesQueuedForProcessing = [];
+
+        const batchesThatDidNotUpload = await this.uploadBatches(
             this.mpInstance.Logger,
-            currentUploads,
+            batchesToUpload,
             useBeacon
         );
-        if (remainingUploads && remainingUploads.length) {
-            this.pendingUploads.unshift(...remainingUploads);
+
+        // Batches that do not successfully upload are added back to the process queue
+        // in the order they were created so that we can attempt re-transmission in
+        // the same sequence. This is to prevent any potential data corruption.
+        if (!isEmpty(batchesThatDidNotUpload)) {
+            // FIXME: This unshift will reverse the order of batches on a retry
+            // this.batchesQueuedForProcessing.unshift(...batchesThatDidNotUpload);
+            this.batchesQueuedForProcessing.push(...batchesThatDidNotUpload);
```

Comment:
I suggest you revert this to unshift to finish this PR, then explain the Sam the situation.  I think this single line coded change should be a PR on its own as a bugfix. This way this PR can then be entitled something like : `refactor: Revise variable names for clarity, add tests for coverage`, and the next one is fixing this bug once you discussed w Sam.  This unblocks you from a release this week.